server/kv.go

```diff
@@ -0,0 +1,694 @@
+	return nil
+}
+
+func (store store) StoreUser(user *User) (returnErr error) {
+	defer func() {
+		if returnErr == nil {
+			return
+		}
+		returnErr = errors.WithMessage(returnErr,
+			fmt.Sprintf("failed to store user, mattermostUserId:%s", user.MattermostUserID))
+	}()
+
+	user.PluginVersion = manifest.Version
+
+	key := hashkey(prefixUser, user.MattermostUserID.String())
+	err := store.set(key, user)
+	if err != nil {
+		return err
+	}
+
+	store.plugin.debugf("Stored: user %s key:%s: connected to:%q", user.MattermostUserID, key, user.ConnectedInstances.IDs())
+	return nil
+}
+
+func (store store) LoadUser(mattermostUserID types.ID) (*User, error) {
+	user := NewUser(mattermostUserID)
+	key := hashkey(prefixUser, mattermostUserID.String())
+	err := store.get(key, user)
+	if err != nil {
+		return nil, errors.WithMessage(err,
+			fmt.Sprintf("failed to load confluence user for mattermostUserId:%s", mattermostUserID))
+	}
+	return user, nil
+}
+
+func (store store) CountUsers() (int, error) {
+	count := 0
+	for i := 0; ; i++ {
+		keys, appErr := store.plugin.API.KVList(i, listPerPage)
+		if appErr != nil {
+			return 0, appErr
+		}
+
+		for _, key := range keys {
+			if strings.HasPrefix(key, prefixUser) {
+				count++
+			}
+		}
+
+		if len(keys) < listPerPage {
+			break
+		}
+	}
+	return count, nil
+}
+
+func (store store) MapUsers(f func(user *User) error) error {
+	for i := 0; ; i++ {
+		keys, appErr := store.plugin.API.KVList(i, listPerPage)
+		if appErr != nil {
+			return appErr
+		}
+
+		for _, key := range keys {
+			if !strings.HasPrefix(key, prefixUser) {
+				continue
+			}
+
+			user := NewUser("")
+			err := store.get(key, user)
+			if err != nil {
+				return errors.WithMessage(err, fmt.Sprintf("failed to load confluence user for key:%s", key))
+			}
+
+			err = f(user)
+			if err != nil {
+				return err
+			}
+		}
+
+		if len(keys) < listPerPage {
+			break
+		}
+	}
+	return nil
+}
+
+// MigrateV2User migrates a user record from the V2 data model if needed. It
+// returns an up-to-date User object either way.
+func (p *Plugin) MigrateV2User(mattermostUserID types.ID) (*User, error) {
+	user, err := p.userStore.LoadUser(mattermostUserID)
+	if errors.Cause(err) != kvstore.ErrNotFound {
+		// return the existing key (or error)
+		return user, err
+	}
+
+	// V3 "user" key does not. Migrate.
+	instances, err := p.instanceStore.LoadInstances()
+	if err != nil {
+		return nil, err
+	}
+
+	user = NewUser(mattermostUserID)
+	for _, instanceID := range instances.IDs() {
+		_, err = p.userStore.LoadConnection(instanceID, mattermostUserID)
+		if errors.Cause(err) == kvstore.ErrNotFound {
+			continue
+		}
+		if err != nil {
+			return nil, err
+		}
+		user.ConnectedInstances.Set(instances.Get(instanceID))
+	}
+	err = p.userStore.StoreUser(user)
+	if err != nil {
+		return nil, err
+	}
+
+	return user, nil
+}
+
+func (store store) StoreOAuth2State(state string) error {
+	// Expire in 15 minutes
```

Comment:
Instead of adding a comment explaining the "magic number", it's always better having a constant defining this number, so the constant name is self-explanatory:

```go
EXPIRY_STORE_TIMEOUT_SECONDS = 15 * 60
```